src/core/vector/qgsvectorlayercache.h

```diff
@@ -417,9 +437,9 @@ class CORE_EXPORT QgsVectorLayerCache : public QObject
 
     void connectJoinedLayers() const;
 
-    inline void cacheFeature( QgsFeature &feat, bool allAttributesFetched )
+    inline void cacheFeature( QgsFeature &feat, bool allAttributesFetched, bool geometryFetched = false )
     {
-      QgsCachedFeature *cachedFeature = new QgsCachedFeature( feat, this, allAttributesFetched );
+      QgsCachedFeature *cachedFeature = new QgsCachedFeature( feat, this, allAttributesFetched, geometryFetched );
```

Comment:
Won't we also need to set the geometryFetched flag when the cache itself ALWAYS requests geometry? Eg something like :
```suggestion
      QgsCachedFeature *cachedFeature = new QgsCachedFeature( feat, this, allAttributesFetched, geometryFetched || mCacheGeometry );
```

Otherwise a call to featureAtIdWithAllAttributesAndGeometry will always re-request a feature even when the original fetch already got the geometry.